ShutdownHook in TracerProviderConfiguration

**Is your feature request related to a problem? Please describe.**
The `TracerProviderConfiguration` registers a ShutdownHook to close the `TracerProvider` (and shutdown the active span processor). In some environments, this may not be the most correct way to do it. Think for instance in a multi-app environment, each with its own `TracerProvider` and with different life cycles. The environments need to call `close` when the app terminates (or undeploys) and not on the JVM shutdown.

**Describe the solution you'd like**
Provide a way to skip the registration of the shutdown hook. This also means that we need to provide a wat to access the `SdkTracerProvider` to call `.close` (or the `Closeable` instance).  Right now this does not seem to be possible if you use auto-configuration.

I have the following issue with the shutdown hook in `TracerProviderConfiguration`: we create a Selenium `RemoteWebDriver` object as singleton bean in Spring Boot, and we want to close/destroy the object only when all tests are done (instead of staring a browser instance before each test and closing it after each test).

The problem is that the bean is closed/destroyed by Spring Boot also using a shutdown hook. Because TracerProviderConfiguration's hook runs faster, the browser app can't be closed anymore, and the Docker selenium/standalone-chrome container keeps having a session until this times out (by default seems to be around 5 minutes).

When the destroy method of the Spring bean is called, we get the following exception:
```
Custom destroy method 'quit' on bean with name 'getWebDriver' threw an exception
org.openqa.selenium.remote.UnreachableBrowserException: Error communicating with the remote browser. It may have died.
Caused by: java.lang.IllegalStateException: Context has not been changed
```

<details>

```
2022-01-05 16:44:55.544 TRACE 20596 --- [ionShutdownHook] o.s.b.f.support.DisposableBeanAdapter    : Invoking custom destroy method 'quit' on bean with name 'getWebDriver'
2022-01-05 16:44:55.567  WARN 20596 --- [ionShutdownHook] o.s.b.f.support.DisposableBeanAdapter    : Custom destroy method 'quit' on bean with name 'getWebDriver' threw an exception

org.openqa.selenium.remote.UnreachableBrowserException: Error communicating with the remote browser. It may have died.
Build info: version: '4.1.1', revision: 'e8fcc2cecf'
System info: host: 'ADAMIAN-T480', ip: '172.28.16.1', os.name: 'Windows 10', os.arch: 'amd64', os.version: '10.0', java.version: '11.0.9.1'
Driver info: org.openqa.selenium.remote.RemoteWebDriver
Command: [dd283e0308b847f856564d6e70159efc, quit {}]
Capabilities {acceptInsecureCerts: false, browserName: chrome, browserVersion: 96.0.4664.110, chrome: {chromedriverVersion: 96.0.4664.45 (76e4c1bb2ab46..., userDataDir: /tmp/.com.google.Chrome.y7HfHN}, goog:chromeOptions: {debuggerAddress: localhost:38217}, javascriptEnabled: true, networkConnectionEnabled: false, pageLoadStrategy: normal, platform: LINUX, platformName: LINUX, proxy: Proxy(), se:cdp: ws://172.17.0.2:4444/sessio..., se:cdpVersion: 96.0.4664.110, se:vnc: ws://172.17.0.2:4444/sessio..., se:vncEnabled: true, se:vncLocalAddress: ws://172.17.0.2:7900, setWindowRect: true, strictFileInteractability: false, timeouts: {implicit: 0, pageLoad: 300000, script: 30000}, unhandledPromptBehavior: dismiss and notify, webauthn:extension:credBlob: true, webauthn:extension:largeBlob: true, webauthn:virtualAuthenticators: true}
Session ID: dd283e0308b847f856564d6e70159efc
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:567) ~[selenium-remote-driver-4.1.1.jar:na]
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:598) ~[selenium-remote-driver-4.1.1.jar:na]
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:602) ~[selenium-remote-driver-4.1.1.jar:na]
	at org.openqa.selenium.remote.RemoteWebDriver.quit(RemoteWebDriver.java:439) ~[selenium-remote-driver-4.1.1.jar:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:566) ~[na:na]
	at org.springframework.beans.factory.support.DisposableBeanAdapter.invokeCustomDestroyMethod(DisposableBeanAdapter.java:318) ~[spring-beans-5.3.14.jar:5.3.14]
	at org.springframework.beans.factory.support.DisposableBeanAdapter.destroy(DisposableBeanAdapter.java:252) ~[spring-beans-5.3.14.jar:5.3.14]
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.destroyBean(DefaultSingletonBeanRegistry.java:587) ~[spring-beans-5.3.14.jar:5.3.14]
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.destroySingleton(DefaultSingletonBeanRegistry.java:559) ~[spring-beans-5.3.14.jar:5.3.14]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.destroySingleton(DefaultListableBeanFactory.java:1161) ~[spring-beans-5.3.14.jar:5.3.14]
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.destroySingletons(DefaultSingletonBeanRegistry.java:520) ~[spring-beans-5.3.14.jar:5.3.14]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.destroySingletons(DefaultListableBeanFactory.java:1154) ~[spring-beans-5.3.14.jar:5.3.14]
	at org.springframework.context.support.AbstractApplicationContext.destroyBeans(AbstractApplicationContext.java:1106) ~[spring-context-5.3.14.jar:5.3.14]
	at org.springframework.context.support.AbstractApplicationContext.doClose(AbstractApplicationContext.java:1075) ~[spring-context-5.3.14.jar:5.3.14]
	at org.springframework.context.support.AbstractApplicationContext.close(AbstractApplicationContext.java:1021) ~[spring-context-5.3.14.jar:5.3.14]
	at org.springframework.boot.SpringApplicationShutdownHook.closeAndWait(SpringApplicationShutdownHook.java:137) ~[spring-boot-2.6.2.jar:2.6.2]
	at java.base/java.lang.Iterable.forEach(Iterable.java:75) ~[na:na]
	at org.springframework.boot.SpringApplicationShutdownHook.run(SpringApplicationShutdownHook.java:106) ~[spring-boot-2.6.2.jar:2.6.2]
	at java.base/java.lang.Thread.run(Thread.java:829) ~[na:na]
Caused by: java.lang.IllegalStateException: Context has not been changed
	at org.openqa.selenium.remote.tracing.opentelemetry.OpenTelemetryContext.createSpan(OpenTelemetryContext.java:62) ~[selenium-remote-driver-4.1.1.jar:na]
	at org.openqa.selenium.remote.tracing.opentelemetry.OpenTelemetryContext.createSpan(OpenTelemetryContext.java:33) ~[selenium-remote-driver-4.1.1.jar:na]
	at org.openqa.selenium.remote.tracing.HttpTracing.newSpanAsChildOf(HttpTracing.java:46) ~[selenium-remote-driver-4.1.1.jar:na]
	at org.openqa.selenium.remote.tracing.TracedHttpClient.execute(TracedHttpClient.java:51) ~[selenium-remote-driver-4.1.1.jar:na]
	at org.openqa.selenium.remote.HttpCommandExecutor.execute(HttpCommandExecutor.java:181) ~[selenium-remote-driver-4.1.1.jar:na]
	at org.openqa.selenium.remote.TracedCommandExecutor.execute(TracedCommandExecutor.java:51) ~[selenium-remote-driver-4.1.1.jar:na]
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:543) ~[selenium-remote-driver-4.1.1.jar:na]
	... 21 common frames omitted

```
</details>

--- a/sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/AutoConfiguredOpenTelemetrySdkBuilder.java
+++ b/sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/AutoConfiguredOpenTelemetrySdkBuilder.java
@@ -16,16 +16,20 @@ import io.opentelemetry.sdk.OpenTelemetrySdkBuilder;
 import io.opentelemetry.sdk.autoconfigure.spi.AutoConfigurationCustomizer;
 import io.opentelemetry.sdk.autoconfigure.spi.AutoConfigurationCustomizerProvider;
 import io.opentelemetry.sdk.autoconfigure.spi.ConfigProperties;
+import io.opentelemetry.sdk.common.CompletableResultCode;
 import io.opentelemetry.sdk.logs.SdkLogEmitterProvider;
 import io.opentelemetry.sdk.metrics.SdkMeterProvider;
 import io.opentelemetry.sdk.resources.Resource;
 import io.opentelemetry.sdk.trace.SdkTracerProvider;
 import io.opentelemetry.sdk.trace.export.SpanExporter;
 import io.opentelemetry.sdk.trace.samplers.Sampler;
+import java.util.ArrayList;
 import java.util.Collections;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
 import java.util.ServiceLoader;
+import java.util.concurrent.TimeUnit;
 import java.util.function.BiFunction;
 import java.util.function.Supplier;
 import java.util.logging.Level;
@@ -57,6 +61,8 @@ public final class AutoConfiguredOpenTelemetrySdkBuilder implements AutoConfigur
 
   private ClassLoader serviceClassLoader =
       AutoConfiguredOpenTelemetrySdkBuilder.class.getClassLoader();
+
+  private boolean registerShutdownHook = true;
 
   private boolean setResultAsGlobal = true;
 
@@ -152,6 +158,22 @@ public final class AutoConfiguredOpenTelemetrySdkBuilder implements AutoConfigur
   }
 
   /**
+   * Control the registration of a shutdown hook to shut down the SDK when appropriate. By default,
+   * the shutdown hook is registered.
+   *
+   * <p>Skipping the registration of the shutdown hook may cause unexpected behavior. This
+   * configuration is for SDK consumers that require control over the SDK lifecycle. In this case,
+   * alternatives must be provided by the SDK consumer to shut down the SDK.
+   *
+   * @param registerShutdownHook a boolean <code>true</code> will register the hook, otherwise
+   *     <code>false</code> will skip registration.
+   */
+  public AutoConfiguredOpenTelemetrySdkBuilder registerShutdownHook(boolean registerShutdownHook) {
+    this.registerShutdownHook = registerShutdownHook;
+    return this;
+  }
+
+  /**
    * Sets whether the configured {@link OpenTelemetrySdk} should be set as the application's
    * {@linkplain io.opentelemetry.api.GlobalOpenTelemetry global} instance.
    */
@@ -200,6 +222,21 @@ public final class AutoConfiguredOpenTelemetrySdkBuilder implements AutoConfigur
     SdkLogEmitterProvider logEmitterProvider =
         LogEmitterProviderConfiguration.configureLogEmitterProvider(
             resource, config, meterProvider);
+
+    if (registerShutdownHook) {
+      Runtime.getRuntime()
+          .addShutdownHook(
+              new Thread(
+                  () -> {
+                    List<CompletableResultCode> shutdown = new ArrayList<>();
+                    shutdown.add(tracerProvider.shutdown());
+                    if (meterProvider instanceof SdkMeterProvider) {
+                      shutdown.add(((SdkMeterProvider) meterProvider).shutdown());
+                    }
+                    shutdown.add(logEmitterProvider.shutdown());
+                    CompletableResultCode.ofAll(shutdown).join(10, TimeUnit.SECONDS);
+                  }));
+    }
 
     ContextPropagators propagators =
         PropagatorConfiguration.configurePropagators(

--- a/sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/LogEmitterProviderConfiguration.java
+++ b/sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/LogEmitterProviderConfiguration.java
@@ -31,9 +31,7 @@ final class LogEmitterProviderConfiguration {
 
     configureLogProcessors(exportersByName).forEach(builder::addLogProcessor);
 
-    SdkLogEmitterProvider logEmitterProvider = builder.build();
-    Runtime.getRuntime().addShutdownHook(new Thread(logEmitterProvider::close));
-    return logEmitterProvider;
+    return builder.build();
   }
 
   // Visible for testing

--- a/sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/MeterProviderConfiguration.java
+++ b/sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/MeterProviderConfiguration.java
@@ -51,12 +51,7 @@ final class MeterProviderConfiguration {
     MetricExporterConfiguration.configureExporter(
         exporterName, config, serviceClassLoader, meterProviderBuilder);
 
-    SdkMeterProvider meterProvider = meterProviderBuilder.build();
-
-    // Make sure metrics shut down when JVM shuts down.
-    Runtime.getRuntime().addShutdownHook(new Thread(meterProvider::close));
-
-    return meterProvider;
+    return meterProviderBuilder.build();
   }
 
   private MeterProviderConfiguration() {}

--- a/sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/TracerProviderConfiguration.java
+++ b/sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/TracerProviderConfiguration.java
@@ -66,9 +66,7 @@ final class TracerProviderConfiguration {
     configureSpanProcessors(config, exportersByName)
         .forEach(tracerProviderBuilder::addSpanProcessor);
 
-    SdkTracerProvider tracerProvider = tracerProviderBuilder.build();
-    Runtime.getRuntime().addShutdownHook(new Thread(tracerProvider::close));
-    return tracerProvider;
+    return tracerProviderBuilder.build();
   }
 
   static List<SpanProcessor> configureSpanProcessors(

